### PR TITLE
FM2-555: Update Medication Dispensing to explicitly set null values a…

### PR DESCRIPTION
**Pull request by mogoodrich**
…s null

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-

## Checklist: I completed these to help reviewers :)

- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java [hunk 1]
```diff
@@ -196,22 +210,38 @@ public MedicationDispense toOpenmrsType(@Nonnull MedicationDispense openmrsObjec
...
 		}
 		
-		if (fhirObject.getMeta() != null) {
-			openmrsObject.setDateCreated(fhirObject.getMeta().getLastUpdated());
+		if (fhirObject.getMeta() != null && fhirObject.getMeta().getLastUpdated() != null) {
+			if (openmrsObject.getDateCreated() == null) {
+				openmrsObject.setDateCreated(fhirObject.getMeta().getLastUpdated());
+			} else {
+				openmrsObject.setDateChanged(fhirObject.getMeta().getLastUpdated());
+			}
```
Comment (mogoodrich) [inline api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java:244]: This is tangential to the ticket, but seemed like the correct behavior... though, perhaps it begs the question, should we really be updating these OpenMRS "audit" fields at all?

**Review by ibacher — APPROVED** [review]
LGTM. Thanks @mogoodrich!

File: api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java (hunk 1 above)
Comment (mogoodrich) [reply api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java:244]: Merging this in, but still curious about thoughts as to whether we should up updated OpenMRS audit fields here... @ibacher  @mseaton

**Comment by mogoodrich** [comment]
Thanks @ibacher , merging!

**Review by mseaton — COMMENTED** [review]
Fine with me, since @ibacher has approved.  Would seem like we should consider having the translators just return null though, if they don't already, and then don't put the "has" check around them.  Then it would just set to null if not found, or set to the appropriate value if found...

**Comment by mogoodrich** [comment]
that's a fair point, @mseaton ... and it may already do that, let me see.

File: api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java (hunk 1 above)
Comment (ibacher) [reply api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java:244]: Oooh... I missed this. I actually don't like the idea of manually updating the audit fields at all... IMO, we can use them to display values when it makes sense, but I don't love allowing anyone to actually edit them.

**Comment by ibacher** [comment]
In general, the translators have been written so that if the supplied value is `null`, they return `null` already. That said, the `has...` check can be more complicated than just a null check, i.e., they usually also call a `<field>.isEmpty()` check which checks for what they regard as an empty element. I don't think it necessarily makes sense for, e.g., the ReferenceTranslator to need to know how to deal with an empty element (since it translates a Reference as a string).

Long story short, I actually think that having these `has...()` checks is the right thing to do because it puts the responsibility for setting the right value on the translator for a particular resource or data element (which should understand the resource or data element) rather than pushing the responsibility to a lower-level translator that may be used for multiple very similar but not identical fields.

File: api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java (hunk 1 above)
Comment (mogoodrich) [reply api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java:244]: thoughts @mseaton?

Comment (mseaton) [reply api-2.6/src/main/java/org/openmrs/module/fhir2/api/translators/impl/MedicationDispenseTranslatorImpl_2_6.java:244]: Fine with me.
